standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java

```diff
@@ -0,0 +1,302 @@
+package org.apache.hadoop.hive.metastore.dbinstall;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hive.metastore.tools.schematool.MetastoreSchemaTool;
+import org.junit.rules.ExternalResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sqlline.SqlLine;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractDatabase extends ExternalResource {
+
+  private final Logger LOG = LoggerFactory.getLogger(AbstractDatabase.class);
+
+  private final int MAX_STARTUP_WAIT = 5 * 60 * 1000;
+  private final String HIVE_USER = "hiveuser";
+  private final String HIVE_PASSWORD = "hivepassword";
+
+  private String db = "hivedb";
+  private boolean verbose = System.getProperty("verbose.schematool") != null;
+  protected boolean useDockerDatabaseArg = false;
+
+  public AbstractDatabase setVerbose(boolean verbose) {
+    this.verbose = verbose;
+    return this;
+  }
+
+  public AbstractDatabase setUseDockerDatabaseArg(boolean useDockerDatabaseArg) {
+    this.useDockerDatabaseArg = useDockerDatabaseArg;
+    return this;
+  }
+
+  public AbstractDatabase setDb(String db) {
+    this.db = db;
+    return this;
+  }
+
+  protected abstract String getDockerImageName();
+  protected abstract List<String> getDockerBaseArgs();
+  protected abstract String getDbType();
+  protected abstract String getDbRootUser();
+  protected abstract String getDbRootPassword();
+  public abstract String getJdbcDriver();
+  protected abstract String getJdbcUrl(String hostAddress);
+  protected abstract boolean isContainerReady(ProcessResults pr);
+
+  public String getJdbcUrl() {
+    return getJdbcUrl(getContainerHostAddress());
+  }
+
+  protected String getDockerDatabaseArg() {
+    return null;
+  }
+
+  /**
+   * URL to use when connecting as root rather than Hive
+   *
+   * @return URL
+   */
+  public abstract String getInitialJdbcUrl(String hostAddress);
+
+  public final String getInitialJdbcUrl() {
+    return getInitialJdbcUrl(getContainerHostAddress());
+  }
+
+  public String getHiveUser(){
+    return HIVE_USER;
+  }
+
+  public String getHivePassword(){
+    return HIVE_PASSWORD;
+  }
+
+  protected String getDb() {
+    return db;
+  }
+
+  private String getDockerContainerName() {
+    return String.format("testDb-%s", getClass().getSimpleName());
+  }
+
+  private List<String> getDockerAdditionalArgs() {
+    List<String> dockerArgs = getDockerBaseArgs();
+    if (useDockerDatabaseArg && StringUtils.isNotEmpty(getDockerDatabaseArg())) {
+      dockerArgs.addAll(Arrays.asList("-e", getDockerDatabaseArg()));
+    }
+    return dockerArgs;
+  }
```

Comment:
Having many `getDocker*Args` method makes the code harder to follow. Does it bring a big benefit having  `getDockerBaseArgs and `getDockerDatabaseArg` separate?

Reply:
Currently every database implementation has a base argument list for docker initialisation, but not all of them has database arguments (eg. Mysql,Oracle). I wanted to make it configurable so we can initialise the database with docker like it was in` AbstractExternalDB.java` or we can create the database in the init/test script like in `DatabaseRule.java`. I wasn’t really sure if I should use one of these approach for both so I kept their original functionality.
What do you think about this?

Reply:
OK makes sense. Thanks for the clarification.